src/v2/cookbook/vue-route-change-handling.md

```diff
@@ -0,0 +1,376 @@
+
+```html
+<router-view></router-view>
+```
+
+and `$watch` the `$route` object:
+
+```js
+Vue.component('OrderIndex', {
+  watch: {
+    $route: {
+      immediate: true,
+      handler: 'loadData',
+    },
+  },
+  methods: {
+    loadData() { /* logic to load data */ },
+  },
+});
+```
+
+<p class="tip">Note! [VueRouter](https://router.vuejs.org/en/) supports [named views](https://router.vuejs.org/en/essentials/named-views.html) and [nested routes](https://router.vuejs.org/en/essentials/nested-routes.html) so the component behaviour depends on the assigned `key` to each `router-view`. It's best to avoid mixing different `key` expressions.</p>
+
+Now each time the route changes, `OrderIndex` reloads its data.
+
+<p class="tip">Remember! Don't modify `$route.query` directly - it does not change the url, use `$router.push` or `$router.replace` instead.</p>
+
+## Simple, but not good enough
+While both `:key` methods work, there are some drawbacks to each one. Lets start with:
+
+```html
+<router-view :key="$route.fullPath"></router-view>
+```
+
+The `$route.fullPath` returns the full url including the query e.g. `/orders?page=2&filter=search`, so what happens when you visit from `/orders` to `/orders?page=1`? The page is recreated with new instances of components and the data is reloaded, yet the route is technically the same as before. At this point the only possible workaround would be to cache the results using something like [VueX](https://vuex.vuejs.org) or plain object. We could use `$route.path` as a `key`, but this will require introducing a watcher and handling query changes which sometimes can be difficult and unexpected side effects if implemented incorrectly. There are more examples why `:key` on the `router-view` could introduce more reloads than necessary which we'll cover further.
+
+Next we have the watch `$route` path without `key` on `router-view`. This gives us more freedom when to reload the data, but it also mean we have to handle it manually. Taking the `/orders` to `/orders?page=1` example, you now can decide if the page query parameter really changed and when to reload the data:
+
+```js
+$route(to, from) {
+  if (to.query.page === from.query.page) {
+    // Page did not change
+    return;
+  } else if (to.query.page && !from.query.page) {
+    // We went from /orders to /order?page=#
+    
+    if (to.query.page === '1') {
+	  // Technically the same page
+	  return;
+    }
+    
+    // We have something like /order?page=2
+    this.loadData();
+  } else {
+    // Our current page and previous page doesn't match
+    this.loadData();
+  }
+}
+```
+
+<p class="tip">Note! All values in `$route.query` are stored as a `string`.</p>
+
+While the snippet could be optimized, it's still quite something just for a `page` query parameter, imagine how it would grow if you have 2 or more similar fields or `page` parameter should reset to `1` when `filter` is changed.
```

Comment:
I'm really not convinced by this whole section 🤔 
Is it about syncing pagination + query with url what you were aiming for?

Reply:
Not really about syncing, but avoiding unnecessary data reloads. As an example, if pagination component does not prevent user from clicking on the already active page button, which updates the url from `/orders` to `/orders?page=1`, the date should not reload as nothing really changed - the user already was on page 1.